app/models/stats/stats_report.rb

```diff
@@ -46,16 +56,28 @@ def self.record_start(report_name)
       create!(report_name: report_name, status: 'started', started_at: Time.now)
     end
 
-    def write_report(report_contents)
-      update(report: report_contents, status: 'completed', completed_at: Time.now)
+    def write_report(report_result)
+      update(
+        document: StringIO.new(report_result.content),
+        document_file_name: "#{report_name}_#{started_at.to_s(:number)}.#{report_result.format}",
+        document_content_type: report_result.content_type,
+        status: 'completed',
+        completed_at: Time.now
+      )
     end
 
     def write_error(report_contents)
       update(report: report_contents, status: 'error', completed_at: nil)
     end
 
+    def document_url(timeout = 10_000)
+      return unless document?
+      document.options[:storage] == :filesystem ? document.path : document.expiring_url(timeout)
+    end
+
     def download_filename
-      "#{report_name}_#{started_at.strftime('%Y_%m_%d_%H_%M_%S')}.csv"
+      # TODO: set the appropriate format as required
```

Comment:
**NOTE:** Once we're using the S3 approach for all reports, this might no longer be required as the document already contains this.